_posts/2019-04-16-accessing-kafka-part-1.md

```diff
@@ -0,0 +1,107 @@
+The only data traffic between the different brokers is the replication, when the follower replicas are receiving their data from the leader replicas.
+That makes the data shards independent on each other and that makes Kafka scale so well.
+
+![Clients connecting to partitions]({{ "/assets/2019-04-16-connecting-to-leader.png" }})
+
+But how do the clients know where to connect?
+
+# Kafka's discovery protocol
+
+Kafka has its own discovery protocol.
+When a Kafka client is connecting to the Kafka cluster, it first connects to any broker which is member of the cluster and asks it for _metadata_ for one or more topics.
+These _metadata_ contain the information about the topics, its partitions and brokers which are responsible for them.
+All brokers should have these data for the whole cluster because they are all synced through Zookeeper.
+Therefore it doesn't matter to which broker the client connected as first - all of them will give it the same response.
+
+![Connection flow between Kafka client and Kafka cluster]({{ "/assets/2019-04-16-connection-flow.png" }})
+
+Once the client gets the _metadata_, it will use them to figure out where to connect when it wants to write to or read from given partition.
+The broker addresses used in the _metadata_ will be either created by the broker it self based on the hostname of the machine where the broker runs.
+Or it can be configured by the used using the `advertised.listeners` option.
+The client will use the address from the _metadata_ to open one or more new connections to the addresses of the brokers which hosts the particular partitions it is interested in.
+And these connection will be used to produce or consume data.
+
+_Note: The description of the Kafka protocol is intentionally simplified for the purpose of this blog post._
+
+# What does it mean for Kafka on Kubernetes
+
+So, what does this mean for running Kafka on Kubernetes?
+If you are familiar with Kubernetes, you probably know that the most common way how to expose some application is using a Kubernetes service.
+Kubernetes services work as _layer 4_ load-balancers.
+They provide a stable DNS address, where the clients can connect.
+And they just forward the connections to one of the pods which are backing the service.
+
+This works reasonably well with most stateless applications which just want to connect randomly to one of the backends behind the service.
+But it can get a lot trickier if your application requires some kind of stickiness.
+It can be for example session stickiness - a client needs to connect to the same pod as last time because of some session information which the pod already has.
+Or it can be a data stickiness - a client needs to connect to a particular pod because it contains some particular data.
+
+This is also the case with Kafka.
+Kubernetes service can be used for the initial connection only - it will take the client to one of the brokers within the cluster where it can get the metadata.
```

Comment:
```suggestion
A Kubernetes service can be used for the initial connection only - it will take the client to _one_ of the brokers within the cluster where it can get the metadata.
```